content/post/difftree.md

```diff
@@ -0,0 +1,501 @@
+first than another with this particular sorting policy.
+
+This path comparison deserve some careful planning thought, as simply comparing
+the string representation of both paths will return wrong results in some corner
+cases.  For instance `a/z.go` should come before `a.go` but the former is
+greater than the later when they are compared as simple strings.  The proper way
+to compare paths is to compare ancestor names between them, going one step at
+a time:
+
+```go
+
+// Compare returns -1, 0 or 1 if the path p comes before, at the same time or
+// after the path o, in lexicographic depth-first order; for example:
+//
+// "a" < "b"
+// "a/b/c/d/z" < "b"
+// "a/b" < "a/b/a"
+// "a/a.go" < "a.go"
+func (p Path) Compare(o Path) int {
+        for i := 0; ; i++ {
+                switch {
+                // there are no more elements in any of the paths
+                case len(o) == len(p) && i == len(p):
+                        return 0
+                // there are no more elements in o
+                case i == len(o):
+                        return 1
+                // there are no more elements in p
+                case i == len(p):
+                        return -1
+                // both paths still have elements, compare the ith one
+                default:
+                        cmp := strings.Compare(p[i].Name(), o[i].Name())
+                        if cmp != 0 {
+                                return cmp
+                        }
+                }
+        }
+}
+```
+
+# The recipie so far
+
+The algorithm described above can be sumarized as:
+
+1. Start traversing both trees at the same time.
+
+2. While there are nodes left to traverse in both trees:
+
+  Compare the names and/or hashes of the elements pointed by both paths and
+  decide:
+
+  - What changes have to be issued, if any.
+
+  - How to advance the walkers: one, the other or both.
+
+3. Issue changes to delete the remaining nodes from A or to insert the
+   remaining nodes from B.
+
+Implementing this algorithm will involve writing:
+
+- A tree iterator that traverses trees in lexicographic depth-first order,
+  returning the path to the nodes.
+
+- A comparator that takes the name and hash of the current paths of two tree
+  iterators and tells us what changes should we issue and how to advance the
+  iterators further.
+
+Let us see how to implement each of them.
+
+# The tree iterator
+
+This will be a regular depth-first tree iterator if it were not for some
+interesting details: 
+
+- siblings are to be iterated in lexicographic order by name.
+
+- it has to be possible to skip whole directories, instead of getting deeper
+  into them.  This is important for efficiency reasons, when the hash of two
+  directories are the same, we want to be able to skip all their contents.
+
+Taking these into account, the iterator type will look something like this:
+
+```go
+type Iterator struct { ... }
```

Comment:
maybe it would be nicer to illustrate this with an interface rather than a struct with methods if the implementations are not going to be shown, what do you think?